storage/netstore.go

```diff
@@ -237,13 +236,11 @@ func (n *NetStore) RemoteFetch(ctx context.Context, req *Request, fi *Fetcher) e
 			log.Trace(err.Error(), "ref", ref)
 			osp.LogFields(olog.String("err", err.Error()))
 			osp.Finish()
-			return ErrNoSuitablePeer
```

Comment:
After setting the loglevel higher you see that the simulation errors with `no peers available` coming from `RemoteFetch`. This is already suspicious since such an error should not reach the user.
With some debugging you can see which `FindPeer` case results in an error. Then grepping the address in the log, you can easily track the series of events: 
- requesting node calls remote fetch which calls `FindPeer` to find a peer 
- the peer starts serving the request
- before delivery, search timeout passes and the following call to `FindPeer` results in an error
- RemoteFetch propagates this error to the NetStore, then downloader, then simulation.

It is not correct to return this error, instead we should return when the request times out.

Reply:
`ErrNoSuitablePeer` should be removed

Reply:
I don't really understand this. If `n.RemoteGet` errors, what are we waiting for... obviously we haven't made any request to remote peers, so?

If we don't care about any errors from `n.RemoteGet`, then we should change the interface. However intuitively if the chunk we are looking for does not exist, why should we hang until the global timeout expires (i.e. the timeout set to the context when the `user` calls `NetStore.Get`?

Reply:
`no peers available` means that all suitable peers have been tried and they did not respond within the `searchTimeout`. if this happens before the `global timeout`, then NetStore gets this error, and not `context deadline exceeded`.

Reply:
Let me try to clarify.
Your line of thinking is based on the assumption that
- requests prompts you to actively look for the chunk
- if these searches are exhausted, you can conclude the chunk is not  found 

However, this is not what your incentive tells you to do. 
-  requestor tells you they require a chunk and willing to pay for it (up to the request TTL, the  proxy for this is currently the request timeout)
-  therefore you really want to be prepared to deliver the chunk to requestor upto TTL expires
and really dont care how it gets there. Simply put, you are not incentivised to close a request before you can no longer earn from the response.

Surely, it does make sense to leave the request open even if you find no further suitable peers to ask. (Note that if there are no suitable peers  to ask for a request, then the node should stop receiving retrieval requests.) This is because there are actually multiple ways the chunk can still appear in our store before the deadline:
 - a peer asked earlier delivers it (later than searchtimeout)
 - new peers are connected which are suitable to ask and they successfully deliver
 - the chunk is pull synced to us as a result of neighbourhood reorganisation
- the chunk is pull synced or push synced to us as a result of it being recently uploaded
- the chunk is actually uploaded by our own node 

This interpretation is much less motivated in traditional server based retrieval, since an asset is either found or not found at the unique place it is meant to be.

Note also that your early abort scheme could be useful if a distinct 404 response would be sent back. However, if this was incentivised, peers could just frivolously respond with it and earn without work. A better way to implement this functionality (tell me if chunk x is currently found in the network) is by asking for proof of custody with a short TTL.)

Hope this helps

Reply:
Thanks for the detailed explanation of your thinking. Maybe I find it hard to wrap my head around since we have very little / close to none of the incentivisation or peers paying or peers willing to pay for requests at the moment.

```
> Surely, it does make sense to leave the request open even if you find no further suitable peers to ask. (Note that if there are no suitable peers to ask for a request, then the node should stop receiving retrieval requests.)
```

If there are no suitable peers for 1 chunk, it doesn't mean that there are no suitable peers for another request. So I am not sure why the node would stop receiving retrieval requests, if they can't seem to find peers that can deliver a particular chunk.

If chunk accounting was part of our tests, and we're optimising for nodes to get the highest amount of currency for chunk deliveries, then this would make sense, but we have none of that.

Reply:
Viktor seems to be thinking about an edge case - the chunk isn't found, but we should not say that it isn't found because there is a small chance that we encounter the chunk _after_ we would have sent a not-found message but _before_ the request times out on its own.

The risk is that this degrades the performance of the entire network (slow/sluggish responses) at minimal gain (if any).

@zelig: The incentive of the storer is to maximise total revenue, not maximise revenue of a particular retrieval request. It only makes sense to get rid of 404 messages and entirely and let every request time-out, if this behaviour doesn't degrade the performance of the network overall to a degree that that there are fewer retrieval requests to serve (fewer users) and thus total income is reduced. 
To answer this we would really have to investigate the relative frequency of the chunk-appearing-just-in-time phenomenon. I suspect it occurs far more frequently in the tests than in real life.